doc/index.rst

```diff
@@ -0,0 +1,600 @@
+
+   See :py:func:`typing.assert_type`. In ``typing`` since 3.11.
+
+   .. versionadded:: 4.2.0
+
+.. function:: clear_overloads()
+
+   See :py:func:`typing.clear_overloads`. In ``typing`` since 3.11.
+
+   .. versionadded:: 4.2.0
+
+.. function:: get_args(tp)
+
+   See :py:func:`typing.get_args`. In ``typing`` since 3.8.
+
+   This function was changed in 3.9 and 3.10 to deal with :data:`Annotated`
+   and :class:`ParamSpec` correctly; ``typing_extensions`` backports these
+   fixes.
+
+.. function:: get_origin(tp)
+
+   See :py:func:`typing.get_origin`. In ``typing`` since 3.8.
+
+   This function was changed in 3.9 and 3.10 to deal with :data:`Annotated`
+   and :class:`ParamSpec` correctly; ``typing_extensions`` backports these
+   fixes.
+
+.. function:: get_original_bases(cls)
+
+   See :py:func:`types.get_original_bases`. Added to the standard library
+   in Python 3.12.
+
+   This function should always produce correct results when called on classes
+   constructed using features from ``typing_extensions``. However, it may
+   produce incorrect results when called on some :py:class:`NamedTuple` or
+   :py:class:`TypedDict` classes on Python <=3.11.
+
+   .. versionadded:: 4.6.0
+
+.. function:: get_overloads(func)
+
+   See :py:func:`typing.get_overloads`. In ``typing`` since 3.11.
+
+   Before Python 3.11, this works only with overloads created through
+   :func:`overload`, not with :py:func:`typing.overload`.
+
+   .. versionadded:: 4.2.0
+
+.. function:: get_type_hints(obj, globalns=None, localns=None, include_extras=False)
+
+   See :py:func:`typing.get_type_hints`.
+
+   In Python 3.11, this function was changed to support the new
+   :py:data:`typing.Required`. ``typing_extensions`` backports these
+   fixes.
+
+   .. versionchanged:: 4.1.0
+
+      Interaction with :data:`Required` and :data:`NotRequired`.
+
+.. function:: is_typeddict(tp)
+
+   See :py:func:`typing.is_typeddict`. In ``typing`` since 3.10.
+
+   On versions where :class:`TypedDict` is not the same as
+   :py:class:`typing.TypedDict`, this function recognizes
+   ``TypedDict`` classes created through either mechanism.
+
+   .. versionadded:: 4.1.0
+
+.. function:: reveal_type(obj)
+
+   See :py:func:`typing.reveal_type`. In ``typing`` since 3.11.
+
+   .. versionadded:: 4.1.0
+
+Other
+~~~~~
+
+.. class:: Text
+
+   See :py:class:`typing.Text`. In ``typing`` since 3.5.2.
+
+.. data:: TYPE_CHECKING
+
+   See :py:data:`typing.TYPE_CHECKING`. In ``typing`` since 3.5.2.
+
+Limitations
+-----------
+
+There are a few types whose interface was modified between different
+versions of typing. For example, :py:class:`typing.Sequence` was modified to
+subclass :py:class:`typing.Reversible` as of Python 3.5.3.
+
+These changes are *not* backported to prevent subtle compatibility
+issues when mixing the differing implementations of modified classes.
+
+Certain types have incorrect runtime behavior due to limitations of older
+versions of the typing module:
```

Comment:
Make it slightly clearer that this is a distinct point to the previous paragraph:

```suggestion
As well as this, certain types have incorrect runtime behavior due to
limitations of older versions of the typing module:
```

Reply:
I actually think we should get rid of this catch-all section and document these limitations only under specific objects. I'll move a version of this text to the ParamSpec docs instead, and get rid of the sentence about Sequence/Reversible which I think is only relevant to 3.5-era changes.

Reply:
Yes, that makes sense. Sounds like a plan.

Reply:
I couldn't figure out what's supposedly wrong with Concatenate, so I'm going to add a note to ParamSpec only.

```
Python 3.8.16 (default, Dec 26 2022, 20:36:56) 
[Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing_extensions import Concatenate, get_args, get_origin, ParamSpec
>>> P = ParamSpec("P")
>>> c = Concatenate[int, P]
>>> get_origin(c)
typing_extensions.Concatenate
>>> get_args(c)
(<class 'int'>, ~P)
```

Reply:
> I couldn't figure out what's supposedly wrong with Concatenate, so I'm going to add a note to ParamSpec only.

Makes sense, though #110 might be worth a mention on the Concatenate entry. I took a brief look at it a while back, and it looks... challenging to backport.

Reply:
#48 as well...

Reply:
Good point, added those. I also noticed that the :gh: links to CPython weren't working, so I added a custom Sphinx extension to make them work (possibly overkill?).

Reply:
> possibly overkill?

Definitely a Choice, but I'll support you in it